tests/test_client/abstract_client.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2019 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use log::error;
+use parsec_client::auth::AuthenticationData;
+use parsec_client::core::basic_client::BasicClient;
+use parsec_client::error::Result;
+use parsec_interface::operations::list_providers::ProviderInfo;
+use parsec_interface::operations::psa_algorithm::{Algorithm, AsymmetricSignature, Hash};
+use parsec_interface::operations::psa_key_attributes::*;
+use parsec_interface::operations::psa_key_attributes::{KeyAttributes, KeyPolicy, UsageFlags};
+use parsec_interface::requests::{Opcode, ProviderID};
+use std::collections::{HashMap, HashSet};
+
+/// Client structure automatically choosing a provider and high-level operation functions.
+#[derive(Debug)]
+pub struct TestClient {
+    op_client: BasicClient,
+    cached_opcodes: Option<HashMap<ProviderID, HashSet<Opcode>>>,
```

Comment:
Is caching opcodes still helpful for our tests? We might only need to know **one** crypto provider ID.

Reply:
> We might only need to know one crypto provider ID.

And isn't that a very good example where provider discovery happens? 😸 You have only one provider in the system so the test client finds it for you.

Reply:
Yes sure, the provider discovery is an excellent idea! I meant since we are doing all requests to one single provider, we might not need to cache this `HashMap` but just be able to do the discovery in the builder and use that as the `provider` field of the BasicClient?